provider/ec2/instancetypes.go

```diff
@@ -22,11 +22,294 @@ import (
+	// when starting new AWS EC2 machines. In this case it's m as the best
+	// general purpose machine type.
+	defaultAWSEC2Family = "m"
+)
+
 var _ environs.InstanceTypesFetcher = (*environ)(nil)
 
+// allInstanceTypeFilter takes a set of filters to run and will only return true
+// for the filter condition if all filters passed. If no filters are supplied
+// then true is returned.
+func allInstanceTypeFilter(filters ...instanceTypeFilter) instanceTypeFilter {
+	return instanceTypeFilterFunc(func(i types.InstanceTypeInfo) bool {
+		for _, f := range filters {
+			if !f.Filter(i) {
+				return false
+			}
+		}
+		return true
+	})
+}
+
+// oneOfInstanceTypeFilter takes a set of filters to run and will return the
+// first true response returned by a filter. If no filters evaluate to true then
+// false is returned.
+func oneOfInstanceTypeFilter(filters ...instanceTypeFilter) instanceTypeFilter {
+	return instanceTypeFilterFunc(func(i types.InstanceTypeInfo) bool {
+		for _, f := range filters {
+			if f.Filter(i) {
+				return true
+			}
+		}
+		return false
+	})
+}
+
+// currentGenInstanceTypeFilter filters out any instance type that is not
+// current generation.
+func currentGenInstanceTypeFilter() instanceTypeFilter {
+	return instanceTypeFilterFunc(func(i types.InstanceTypeInfo) bool {
+		return aws.ToBool(i.CurrentGeneration)
+	})
+}
+
+// exactInstanceTypeFilter filters out instances that don't exactly match the
+// instance type supplied.
+func exactInstanceTypeFilter(match types.InstanceType) instanceTypeFilter {
+	return instanceTypeFilterFunc(func(i types.InstanceTypeInfo) bool {
+		return i.InstanceType == match
+	})
+}
+
+// selectorInstanceTypeFilter will match on instances that contain the same
+// values specified in the instance types. Zero values will be ignored.
+func selectorInstanceTypeFilter(selector instanceType) instanceTypeFilter {
+	return instanceTypeFilterFunc(func(i types.InstanceTypeInfo) bool {
+		itDetails, err := parseInstanceType(i.InstanceType)
+		if err != nil {
+			return false
+		}
+		if !itDetails.Capabilities.Difference(selector.Capabilities).IsEmpty() {
+			return false
+		}
+		if itDetails.Family != selector.Family && selector.Family != "" {
+			return false
+		}
+		if itDetails.ProcessorFamily != selector.ProcessorFamily && selector.ProcessorFamily != "" {
+			return false
+		}
+		if itDetails.Generation != selector.Generation && selector.Generation != 0 {
+			return false
+		}
+		if itDetails.Size != selector.Size && selector.Size != "" {
+			return false
+		}
+
+		return true
+	})
+}
+
+// generalPurposeInstanceFilter supplies a filter that is capable of filtering
+// only on machines that are considered general purpose enough for Juju to use
+// as a sane default.
+func generalPurposeInstanceFilter(
+	ctx context.ProviderCallContext,
+	cache *instanceTypeCache,
+) (instanceTypeFilter, error) {
+	highestGenerationIntel, err := cache.HighestFamilyGeneration(ctx, defaultAWSEC2Family+processorFamilyIntel)
```

Comment:
we should pass family and processor type as seperate arguments, as the key for the HighestFamilyGeneration map is an implementation detail of the cache.